element_volume/volume.py

```diff
@@ -0,0 +1,261 @@
+            TIF file of the volumetric data associated with a given scan.
+    """
+
+    if isinstance(linking_module, str):
+        linking_module = importlib.import_module(linking_module)
+    assert inspect.ismodule(
+        linking_module
+    ), "The argument 'linking_module' must be a module's name or a module"
+
+    global _linking_module
+    _linking_module = linking_module
+
+    schema.activate(
+        schema_name,
+        create_schema=create_schema,
+        create_tables=create_tables,
+        add_objects=_linking_module.__dict__,
+    )
+
+
+# -------------------------- Functions required by the Element -------------------------
+
+
+def get_volume_root_data_dir() -> list:
+    """Fetches absolute data path to ephys data directories.
+
+    The absolute path here is used as a reference for all downstream relative paths used in DataJoint.
+
+    Returns:
+        A list of the absolute path(s) to volume data directories.
+    """
+    root_directories = _linking_module.get_vol_root_data_dir()
+    if isinstance(root_directories, (str, Path)):
+        root_directories = [root_directories]
+
+    return root_directories
+
+
+def get_volume_tif_file(scan_key: dict) -> (str, Path):
+    """Retrieve the full path to the TIF file of the volumetric data associated with a given scan.
+    Args:
+        scan_key: Primary key of a Scan entry.
+    Returns:
+        Full path to the TIF file of the volumetric data (Path or str).
+    """
+    return _linking_module.get_volume_tif_file(scan_key)
+
+
+# --------------------------------------- Schema ---------------------------------------
+
+@schema
+class Volume(dj.Imported):
+    definition = """
+    -> Scan
+    ---
+    px_width: int # total number of voxels in x dimension
+    px_height: int # total number of voxels in y dimension
+    px_depth: int # total number of voxels in z dimension
+    depth_mean_brightness=null: longblob  # mean brightness of each slice across the depth (z) dimension of the stack
+    volume: longblob  # volumetric data - np.ndarray with shape (z, y, x)
+    """
+
+    def make(self, key):
+        vol_tif_fp = get_volume_tif_file(key)
+        volume_data = TiffFile(vol_tif_fp).asarray()
+
+        self.insert1(
+            dict(
+                **key,
+                volume=volume_data,
+                px_width=volume_data.shape[2],
+                px_height=volume_data.shape[1],
+                px_depth=volume_data.shape[0]
+            )
+        )
+
+
+@schema
+class SegmentationParamset(dj.Lookup):
+    definition = """
+    paramset_idx: int
+    ---
+    segmentation_method: varchar(32)
+    paramset_desc="": varchar(256)
+    params: longblob
+    paramset_hash: uuid
+    unique index (paramset_hash)
+    """
+
+    @classmethod
+    def insert_new_params(
+        cls,
+        segmentation_method: str,
+        paramset_desc: str = "",
+        params: dict = {},
```

Comment:
I believe that `params` should not be optional.

Reply:
fixed